File: leet/two_pointer/move_zeroes_283.py
class Solution(object):
    def moveZeroes(self, nums):
        zero_list = []
        c_list = []
        
        for i in range(len(nums)):
            zero_list.append(0) if nums[i]==0 else c_list.append(nums[i])                
             
        nums.clear()
        c_list.extend(zero_list)
        nums.extend(c_list)


aa = Solution()
aa.moveZeroes([0,1,0,3,12])


# c_list.pop(i) 
# 不能直接在複製的list推掉東西，會修改到這複製品的長度
# ! 這邊也可以用two pointer作

File: leet/two_pointer/move_zeroes_283.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public void moveZeroes(int[] nums) {
        int dot = 0;
        for (int i : nums) {
            if (i != 0) {
                nums[dot] = i;
                dot++;
            }
        }
        while (dot < nums.length) {
            nums[dot] = 0;
            dot++;
        }
    }
}

// ----------------------------------------------------------------------
// !nums 這邊是primitive，所以clear()這函數不能用
// class Solution {
// public void moveZeroes(int[] nums) {
// List<Integer> zeroList = new ArrayList<>();
// List<Integer> cList = new ArrayList<>();

// for (int i = 0; i < nums.length; i++) {
// if (nums[i] == 0) {
// zeroList.add(0);
// } else {
// cList.add(nums[i]);
// }
// }
// ! nums.clear();
// cList.addAll(zeroList);
// nums.addAll(cList);
// }
// }